File: android-app/app/src/main/java/lvt/crm/AppContainer.kt
package lvt.crm

import android.content.Context
import lvt.crm.data.auth.AuthRepository
import lvt.crm.data.auth.TokenStore
import lvt.crm.data.convex.ConvexConfig
import lvt.crm.data.convex.ConvexHttpClient
import lvt.crm.data.duties.DutiesRepository
import lvt.crm.data.notifications.NotificationsRepository
import lvt.crm.data.work.WorkRepository
import lvt.crm.push.NotificationScheduler
import lvt.crm.push.FcmTokenRegistrar

class AppContainer(context: Context) {
    val appContext = context.applicationContext
    val tokenStore = TokenStore(appContext)

    val convex = ConvexHttpClient(
        baseUrl = ConvexConfig.url,
        tokenProvider = { tokenStore.accessToken },
        refreshCredentialsProvider = { tokenStore.snapshot() },
        onTokensRefreshed = { expected, access, refresh ->
            tokenStore.replaceIfCurrent(expected, access, refresh)
        },
    )

    val fcmTokenRegistrar = FcmTokenRegistrar(appContext, tokenStore, convex)
    val authRepository = AuthRepository(
        tokenStore,
        convex,
        beforeSignOut = { accessToken -> fcmTokenRegistrar.unregister(accessToken) },
    )
    val dutiesRepository = DutiesRepository(convex)
    val notificationsRepository = NotificationsRepository(convex)
    val workRepository = WorkRepository(convex)
    val notificationScheduler = NotificationScheduler(appContext)
}

File: ios-uikit-lvt/LvtCrmUIKit/AppContainer.swift
import Foundation

@MainActor
final class AppContainer {
    let tokenStore: TokenStore
    let convex: ConvexHttpClient
    let authRepository: AuthRepository
    let sessionsRepository: SessionsRepository
    let notificationsRepository: NotificationsRepository
    let dutiesRepository: DutiesRepository
    let workRepository: WorkRepository
    let notificationSync: NotificationSyncService
    let apnsRegistrar: APNsTokenRegistrar

    init() {
        let tokenStore = TokenStore()
        self.tokenStore = tokenStore
        let tokenProvider: @Sendable () -> String? = { tokenStore.accessToken }
        let refreshCredentialsProvider: @Sendable () -> CredentialSnapshot? = { tokenStore.snapshot() }
        let onTokensRefreshed: @Sendable (CredentialSnapshot, String, String) -> Bool = { expected, access, refresh in
            tokenStore.replaceIfCurrent(expected, accessToken: access, refreshToken: refresh)
        }
        let convex = ConvexHttpClient(
            baseURL: ConvexConfig.url,
            tokenProvider: tokenProvider,
            refreshCredentialsProvider: refreshCredentialsProvider,
            onTokensRefreshed: onTokensRefreshed
        )
        self.convex = convex
        sessionsRepository = SessionsRepository(convex: convex)
        let notificationsRepository = NotificationsRepository(convex: convex)
        self.notificationsRepository = notificationsRepository
        dutiesRepository = DutiesRepository(convex: convex, tokenProvider: { tokenStore.accessToken })
        workRepository = WorkRepository(convex: convex, tokenProvider: { tokenStore.accessToken })
        notificationSync = NotificationSyncService(
            tokenStore: tokenStore,
            repository: notificationsRepository
        )
        let apnsRegistrar = APNsTokenRegistrar(tokenStore: tokenStore, convex: convex)
        self.apnsRegistrar = apnsRegistrar
        authRepository = AuthRepository(
            tokenStore: tokenStore,
            convex: convex,
            beforeSignOut: { accessToken in
                await apnsRegistrar.unregister(accessToken: accessToken)
            }
        )
    }
}
